src/bucket/BucketManagerImpl.cpp

```diff
@@ -867,4 +868,165 @@ BucketManagerImpl::isShutdown() const
+                if (e.type() == LIVEENTRY || e.type() == INITENTRY)
+                {
+                    (*map)[LedgerEntryKey(e.liveEntry())] = e.liveEntry();
+                }
+                else
+                {
+                    if (e.type() != DEADENTRY)
+                    {
+                        throw std::runtime_error("Malformed bucket: unexpected "
+                                                 "non-INIT/LIVE/DEAD entry.");
+                    }
+                    map->erase(e.deadEntry());
+                }
+            }
+            ++in;
+        }
+    });
+    double mb = static_cast<double>(sz) / 1000000.0;
+    double ms = timer.max();
+    double sec = ms / 1000.0;
+    CLOG_INFO(Bucket, "{} {}-byte bucket file in {:.2f}ms ({:.2f} MB/s)", pass,
+              sz, ms, mb / sec);
+}
+
+bool
+BucketManagerImpl::selfCheck(
+    HistoryArchiveState const& has,
+    UnorderedMap<LedgerKey, LedgerEntry>& ledgerMap) const
+{
+    bool ok = true;
+
+    std::vector<std::string> files =
+        fs::findfiles(getBucketDir(), isBucketFile);
+    std::set<Hash> hashesInDir;
+    for (auto const& f : files)
+    {
+        hashesInDir.insert(extractFromFilename(f));
+    }
+
+    // First we collect the set and order of buckets we are expecting to see.
+    std::set<Hash> hashesInBl;
+    std::set<Hash> hashesInFuture;
+    std::vector<Hash> hashOrderToVisit;
+    for (uint32_t i = BucketList::kNumLevels; i > 0; --i)
+    {
+        HistoryStateBucket const& hsb = has.currentBuckets.at(i - 1);
+        Hash snap = hexToBin256(hsb.snap);
+        Hash curr = hexToBin256(hsb.curr);
+        hashesInBl.insert(snap);
+        hashesInBl.insert(curr);
+        std::vector<Hash> expected = {snap, curr};
+        if (hsb.next.hasOutputHash())
+        {
+            Hash next = hexToBin256(hsb.next.getOutputHash());
+            hashesInFuture.insert(next);
+            expected.emplace_back(next);
+        }
+        for (auto const& e : expected)
+        {
+            if (isZero(e))
+            {
+                continue;
+            }
+            hashOrderToVisit.emplace_back(e);
+            if (hashesInDir.find(e) == hashesInDir.end())
+            {
+                CLOG_ERROR(Bucket,
+                           "Missing expected level {} file for bucket {}",
+                           i - 1, binToHex(e));
+                ok = false;
+            }
+            else
+            {
+                CLOG_INFO(Bucket, "Found expected level {} file for bucket {}",
+                          i - 1, binToHex(e));
+            }
+        }
+    }
+
+    for (auto const& h : hashesInDir)
+    {
+        if (hashesInBl.find(h) == hashesInBl.end() &&
+            hashesInFuture.find(h) == hashesInFuture.end())
+        {
+            // This is a hash not-in the BL or a FutureBucket. Warn, but not
+            // error.
+            CLOG_WARNING(Bucket, "found unexpected bucket {}", binToHex(h));
```

Comment:
Well, it's only a warning -- nothing fatal, just noteworthy.

Since self-check is run when core is offline and `DISABLE_BUCKET_GC` is an abnormal setting, I think there shouldn't be any reason for stale (or otherwise unexpected) buckets most of the time; I think it's worth raising the issue as it indicates a _possible_ logic, configuration or operator error somewhere.

Reply:
> I think there shouldn't be any reason for stale (or otherwise unexpected) buckets most of the time;

I don't think we can guarantee this, unfortunately. The problem is that the current bucketlist is often a subset of the buckets core is actually using. For example, we reference stale buckets in the publish queue, so with this check those will be reported as "unexpected", even though it's perfectly normal to retain those while we're publishing. 

I am also not sure we can definitively say that core doesn't leave any unreferenced buckets in the buckets folder, I recall seeing unreferenced buckets that just get cleaned up on restart (although this was a while ago, so maybe it got fixed. But I don't think we ever did a complete audit to ensure no unreferenced buckets are left). So I think it's better to be conservative here and not give the user potentially misleading warnings.

Reply:
Well .. I can (and will change this to) account for the referenced buckets in the publish queue by excluding those covered by getReferencedBuckets() as well. The odds of this are low anyways -- something like 95% of the time the publish queue is empty -- but sure, it's possible. Worth excluding.

Besides that and merges-in-progress (which aren't active here -- we didn't start the app / call assumeState) I think there should not be any others. If there are, I'm curious to know!